File: Java/ds/Stack.java
import java.util.*;
public class Stack
{
  static Node top;
  class Node
  {
	public int data;
	public Node next;
	Node(int data)
	{
		this.data=data;
	}
  }
  void push(int data)
  {
	Node ptr = new Node(data);
	if(top==null)
	{
		top=ptr;
		ptr.next=null;
	}
	else 
	{
		Node ctr=top;
		ptr.next=ctr;
		top = ptr;
	}
  }
  boolean isEmpty()
  {
	if(top==null) 
		return true;
	return false;
  }
  void peek()
  {     if(!isEmpty())
	  System.out.println(top.data);
        else 
          System.out.println("Stack is empty");
  }
  void pop()
  {
	Node ctr=top;
	if(ctr==null)
	{
		System.out.println("Stack is empty");
		return ;
	}
	else if(ctr!=null && ctr.next==null)
		top = null;
	else 
		top = top.next;	
  }
  void print()
  {
	Node ctr=top;
	while(ctr!=null)
	{
		System.out.print(ctr.data+" ");
		ctr=ctr.next;
	}
	System.out.println();
  }
  public static void main(String[] args)
  {
	Stack st=new Stack();
	Scanner sc=new Scanner(System.in);
	System.out.println("Stack");
	int choice; boolean f=true;
	while(f) 
	{
		System.out.println("Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit");
		choice=sc.nextInt();
		switch(choice) 
		{
		 case 1: System.out.println("Enter data to enter into stack");
		         int data = sc.nextInt();
		         st.push(data);
		         st.print();
		         System.out.println("Success");
		         break;
		 case 2: st.pop();
		         st.print();
		         System.out.println("Success");
			 break;
		 case 3: st.peek();
			  System.out.println("Success");	
		         break;
		case 4:  f=false;
	                 break;
	     default: System.out.println("Enter Valid Input");
	            break;
		}
	}
  }

}

/*
Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit
1
Enter data to enter into stack
2
2 
Success
Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit
1
Enter data to enter into stack
3
3 2 
Success
Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit
2
2 
Success
Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit
3
2
Success
Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit
1
Enter data to enter into stack
4
4 2 
Success
Enter 1 to input data, 2 to delete data from stack, 3 to peek data from stack and 4 to exit
4
*/

File: Go/ds/Stack.go
package main

import (
	"errors"
	"fmt"
	"math/rand"
)

// A linked list node
type node struct {
	data interface{}
	next *node
}

type stack struct {
	top  *node
	size int
}

func (obj *stack) isEmpty() bool {
	return obj.size == 0
}

// Adds the given element to the beginning of the stack
func (obj *stack) Push(value interface{}) {
	obj.top = &node{
		data: value,
		next: obj.top,
	}
	obj.size++
}

// Removes the element at the beginning of the stack
// and returns the element along with an error message
// if the stack is empty else 'nil'
func (obj *stack) Pop() (interface{}, error) {
	if obj.top == nil {
		return nil, errors.New("Empty stack")
	}

	var data interface{}
	data, obj.top = obj.top.data, obj.top.next
	obj.size--

	return data, nil
}

// Returns a random element to insert in the Stack
func getElement() interface{} {
	var temp interface{} = rand.Intn(100)
	fmt.Println("Element: ", temp)

	return temp
}

// Returns the value at the beginning of the stack
// along with the error message
func (obj *stack) Peek() (interface{}, error) {
	if obj.top == nil {
		return nil, errors.New("Empty stack")
	}

	return obj.top.data, nil
}

func main() {
	var object stack
	var flag bool = true
	var choice int

	for flag {
		fmt.Print("Menu\n1. Push element\n2. Pop element\n3. Peek element\n4. Count of elements\n5. Exit\nEnter your choice: ")
		fmt.Scan(&choice)

		switch choice {
		case 1:
			element := getElement()
			object.Push(element)
		case 2:
			element, err := object.Pop()
			fmt.Println("Element: ", element)
			fmt.Println("Error: ", err)
		case 3:
			element, err := object.Peek()
			fmt.Println("Element: ", element)
			fmt.Println("Error: ", err)
		case 4:
			fmt.Println("The total number of elements present in the Stack are: ", object.size)
		case 5:
			flag = false
		default:
			fmt.Println("Please enter valid choice.")
		}
	}
}
